app/spdb/scripts/spdb

```diff
@@ -0,0 +1,256 @@
+
+# psjava access.
+export PATH=$PATH:/home/software/bin
+
+# System variables.
+SERVER=`hostname`
+PORT=8224
+
+
+# Display usage information.
+usage ()
+{
+    echo $"Usage: $0 {start|stop|status|restart}" 1>&2
+    RETVAL=1
+}
+
+# Sleep for 10 seconds, and display a . after each two second interval.
+sleep_dots ()
+{
+    for i in `seq 1 5`; do
+	printf " ."
+	sleep 2
+    done
+}
+
+# Find the latest version of the ODB.
+# If one is found:
+# 1. SPDB_DIR is set to the path of the spdb directory.
+# 2. SPDB_VERSION is set to the version string (e.g. 1.2.3).
+# 4. SPDB_EXEC is set to the path to the binary.
+determine_spdb_directory ()
+{
+    local SPDB_DIRS=($(ls -d /home/software/spdb*))
+    unset SPDB_DIR
+    unset SPDB_LOCKFILE
+    unset SPDB_VERSION_STR
+    SPDB_SEMESTER=0000A
+    SPDB_VERSION=(0 0 0)
+    
+    for CURR_DIR in "${SPDB_DIRS[@]}"
+    do
+	# Extract the version so that we can find the most recent.
+	# We don't use sort on the ls output for this because it will compare by string
+	# instead of by number, and we want to avoid things like 1.10.1 being less than 1.2.1.
+	CURR_SEMESTER=`echo $CURR_DIR | sed "s/^.*spdb_\([^-.]*\).*$/\1/"`
+	if [[ "$CURR_SEMESTER" < "$SPDB_SEMESTER" ]]; then
+	    continue
+	fi
+	CURR_VERSION_STR=`echo $CURR_DIR | sed "s/^[^.]*\.\([^_]*\).*$/\1/"`
+	if [[ -z "$CURR_VERSION_STR" ]]; then
+	    continue
+	fi
+	
+	CURR_VERSION=(${CURR_VERSION_STR//./ })
+	if [[ -z "${CURR_VERSION[@]}" ]]; then
+	    continue
+	fi
+	if [[ "${#CURR_VERSION[@]}" -ne 3 ]]; then
+	    continue
+	fi
+	
+	# Now compare versions. If the new version number is no greater than the previous, disregard.
+	if [[ "$CURR_SEMESTER" == "$SPDB_SEMESTER" ]]; then
+	    if [[ "${SPDB_VERSION[0]}" -gt "${CURR_VERSION[0]}" ]]; then
+		continue
+	    elif [[ "${SPDB_VERSION[0]}" -eq "${CURR_VERSION[0]}" ]]; then
+		if [[ "${SPDB_VERSION[1]}" -gt "${CURR_VERSION[1]}" ]]; then
+		    continue
+		elif [[ "${SPDB_VERSION[1]}" -eq "${CURR_VERSION[1]}" ]]; then
+		    if [[ "${SPDB_VERSION[2]}" -gt "${CURR_VERSION[2]}" ]]; then
+			continue
+		    fi
+		fi
+	    fi
+	fi
+	
+	# We have encountered a newer version.
+	SPDB_SEMESTER=$CURR_SEMESTER
+	SPDB_VERSION_STR=$CURR_VERSION_STR
+	SPDB_VERSION=(${CURR_VERSION[@]})
+	SPDB_DIR=$CURR_DIR
+    done
+
+    if [[ -z "$SPDB_DIR" ]]; then
+	echo "Could not find most recent ODB installation. Aborting."
+	return 1
+    fi
+
+    # Set the SPDB executable and the lockfile.
+    SPDB_EXEC=`ls ${SPDB_DIR}/spdb* 2>/dev/null`
+    SPDB_LOCKFILE=`ls /home/software/.ocs15/spdb_${SPDB_SEMESTER}*${SPDB_VERSION_STR}*/lockfile 2>/dev/null`
+    return 0
+}
+
+
+# If there is already a running SPDB:
+# 1. Determine its PID and set SPDB_PID to it.
+# 2. Determine the version (e.g. 2018A-test.1.6.1) and set SPDB_VER to it.
+# Returns 0 if running, and nonzero (1) otherwise.
+find_running_spdb ()
+{
+    SPDB_PID=`psjava | awk '{ printf $2; }'`
+    SPDB_VER=`psjava | sed "s/^.*\(spdb_[^_]*\).*/\1/"`
```

Comment:
We're assuming that the only java process running will be the `spdb`.  I think that's true for now but would silently break if we ever run another java process?

Reply:
This is a good point. We aren't great at cleaning up old installations. I wrote this so that it could operate on all ODB machines, including `gsodbtest` and `gnodbtest`, where I must admit that I let old versions linger too long. I'll be more diligent about this and do some cleanup in preparation for the release.

You are right about the `psjava` issue. I may as well make it a `psspdb` call that filters only for spdb.